radio/src/timers.cpp

```diff
@@ -0,0 +1,209 @@
+ * gruvin9x by Bryan J. Rentoul: http://code.google.com/p/gruvin9x/,
+ * er9x by Erez Raviv: http://code.google.com/p/er9x/,
+ * and the original (and ongoing) project by
+ * Thomas Husterer, th9x: http://code.google.com/p/th9x/
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ */
+
+#include "opentx.h"
+#include "timers.h"
+
+TimerState timersStates[MAX_TIMERS] = { { 0 }, { 0 } };
+
+
+void timerReset(uint8_t idx)
+{
+  TimerState & timerState = timersStates[idx];
+  timerState.state = TMR_OFF; // is changed to RUNNING dep from mode
+  timerState.val = g_model.timers[idx].start;
+  timerState.val_10ms = 0 ;
+}
+
+#if defined(CPUARM)
+void timerSet(int idx, int16_t val)
+{
+  TimerState & timerState = timersStates[idx];
+  timerState.state = TMR_OFF; // is changed to RUNNING dep from mode
+  timerState.val = val;
+  timerState.val_10ms = 0 ;
+}
+#endif // #if defined(CPUARM)
+
+#if defined(CPUARM) || defined(CPUM2560)
+void restoreTimers()
+{
+  for (uint8_t i=0; i<MAX_TIMERS; i++) {
+    if (g_model.timers[i].persistent) {
+      timersStates[i].val = g_model.timers[i].value;
+    }
+  }
+}
+
+void saveTimers()
+{
+  for (uint8_t i=0; i<MAX_TIMERS; i++) {
+    if (g_model.timers[i].persistent) {
+      TimerState *timerState = &timersStates[i];
+      if (g_model.timers[i].value != (uint16_t)timerState->val) {
+        g_model.timers[i].value = timerState->val;
+        eeDirty(EE_MODEL);
+      }
+    }
+  }
+
+#if defined(CPUARM) && !defined(REVA)
+  if (sessionTimer > 0) {
+    g_eeGeneral.globalTimer += sessionTimer;
+    eeDirty(EE_GENERAL);
+    sessionTimer = 0;
+  }
+#endif
+}
+#endif  // #if defined(CPUARM) || defined(CPUM2560)
+
+#if defined(ACCURAT_THROTTLE_TIMER)
+  #define THR_TRG_TRESHOLD    13      // approximately 10% full throttle
+#else
+  #define THR_TRG_TRESHOLD    3       // approximately 10% full throttle
+#endif
+
+void evalTimers(int16_t throttle, uint8_t tick10ms)
+{
+  for (uint8_t i=0; i<MAX_TIMERS; i++) {
+    int8_t tm = g_model.timers[i].mode;
+    uint16_t tv = g_model.timers[i].start;
+    TimerState * timerState = &timersStates[i];
+
+    if (tm) {
+      if ((timerState->state == TMR_OFF) && (tm != TMRMODE_THR_TRG)) {
+        timerState->state = TMR_RUNNING;
+        timerState->cnt = 0;
+        timerState->sum = 0;
+      }
+
+      if (tm == TMRMODE_THR_REL) {
+        timerState->cnt++;
+        timerState->sum += throttle;
+      }
+
+      if ((timerState->val_10ms += tick10ms) >= 100) {
+        if (timerState->val == TIMER_MAX) break;
+        if (timerState->val == TIMER_MIN) break;
```

Comment:
Added timer wrap-arround stopper. This was another (unrelated) issue, but I decided to fix it. It increases stock FLASH size.